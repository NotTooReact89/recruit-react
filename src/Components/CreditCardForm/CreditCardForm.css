.ccForm {
	width: 500px;
	margin: auto;
	padding: 1.5rem;
}

.btn-custom {
	background-color: #fcbd1b;
	color: #000000
}

.input-custom {
	background-color: #D6D6D6;
	color: #555;
}