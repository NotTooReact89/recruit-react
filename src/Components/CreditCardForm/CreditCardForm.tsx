import * as React from 'react';
import Log from '../../Lib/Log';
import './CreditCardForm.css';
import { FormErrors } from './FormErrors';

interface IFormError {
  ccNumber: string,
  cvvNumber: string,
  expiryDate: string
}

interface IFormState {
  ccNumber: string,
  cvvNumber: string,
  ccNumberValid: boolean,
  cvvNumberValid: boolean,
  expiryDate: string,
  expiryDateValid: boolean,
  formErrors: IFormError,
  formValid: boolean
}
class CreditCardForm extends React.Component<any, IFormState> {
  constructor (props: any) {
    super(props);
    this.state = {
      ccNumber: '',
      ccNumberValid: false,
      cvvNumber: '',
      cvvNumberValid: false,
      expiryDate: '',
      expiryDateValid: false,
      formErrors: {ccNumber: '', cvvNumber: '', expiryDate: ''},
      formValid: false
    }
  }

  public handleUserInput = (e: any) => {
    const name = e.target.name;
    const value = e.target.value;
    this.setState({[name]: value} as any,
                  () => { this.validateField(name, value) });
  }

  public validateField(fieldName: string, value: string) {
    const fieldValidationErrors = this.state.formErrors;
    let ccNumberValid = this.state.ccNumberValid;
    let cvvNumberValid = this.state.cvvNumberValid;
    let expiryDateValid = this.state.expiryDateValid;

    switch(fieldName) {
      case 'ccNumber':
      ccNumberValid = value.length >= 1;
        fieldValidationErrors.ccNumber = ccNumberValid ? '' : ' should not be empty';
        break;
      case 'cvvNumber':
        cvvNumberValid = value.length >= 1;
        fieldValidationErrors.cvvNumber = cvvNumberValid ? '': ' should not be empty';
        break;
      case 'expiryDate':
        expiryDateValid = value.length >= 1;
        fieldValidationErrors.expiryDate = expiryDateValid ? '': ' should not be empty';
        break;
      default:
        break;
    }
    this.setState({
      ccNumberValid,
      cvvNumberValid,
      expiryDateValid,
      formErrors: fieldValidationErrors
    }, this.validateForm);
  }

  public validateForm() {
    this.setState({formValid: this.state.ccNumberValid && this.state.cvvNumberValid && this.state.expiryDateValid});
  }

  public errorClass(error: string) {
    return(error.length === 0 ? '' : 'has-error');
  }

  public handleSubmit = (event: any) => {
    event.preventDefault()
    if (this.state.formValid) {
        Log.info(`Credit card number: ${event.target.ccNumber.value}`)
        Log.info(`CVV number: ${event.target.cvvNumber.value}`)
        Log.info(`Expiry number: ${event.target.expiryDate.value}`)
    }
  }

  public render () {
    return (
      <form onSubmit={this.handleSubmit} className="ccForm">
        <h2>Enter your credit card details</h2>
        <div className="panel panel-default">
          <FormErrors formErrors={this.state.formErrors} />
        </div>
        <div className={`form-group ${this.errorClass(this.state.formErrors.ccNumber)}`}>
          <label htmlFor="ccNumber">Credit card number</label>
          <input type="text" required={true} className="form-control input-custom" name="ccNumber"
            placeholder="Your 10 digit credit card number"
            value={this.state.ccNumber}
            onChange={this.handleUserInput}  />
        </div>
        <div className={`form-group ${this.errorClass(this.state.formErrors.cvvNumber)}`}>
          <label htmlFor="cvvNumber">CVV number</label>
          <input type="password" required={true} className="form-control input-custom" name="cvvNumber"
            placeholder="Three digit at the back of your card"
            value={this.state.cvvNumber}
            onChange={this.handleUserInput}  />
        </div>
        <div className={`form-group ${this.errorClass(this.state.formErrors.expiryDate)}`}>
          <label htmlFor="expiryDate">Expiry date</label>
          <input type="date" required={true} className="form-control input-custom" name="expiryDate"
            placeholder="Your credit card expiry date"
            value={this.state.expiryDate}
            onChange={this.handleUserInput}  />
        </div>
        <button type="submit" className="btn btn-custom">Submit</button>
      </form>
    )
  }
}

export default CreditCardForm;