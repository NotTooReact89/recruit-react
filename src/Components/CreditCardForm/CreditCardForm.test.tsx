import * as React from 'react';
import CreditCardForm from './CreditCardForm';
const renderer = require('react-test-renderer');

it('should render credit card form component correctly', () => {
  const tree = renderer.create((
    <CreditCardForm />
  )).toJSON()
  expect(tree).toMatchSnapshot()
})