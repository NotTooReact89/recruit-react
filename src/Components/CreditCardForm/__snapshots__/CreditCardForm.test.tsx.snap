// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render credit card form component correctly 1`] = `
<form
  className="ccForm"
  onSubmit={[Function]}
>
  <h2>
    Enter your credit card details
  </h2>
  <div
    className="panel panel-default"
  >
    <div
      className="formErrors"
    >
      
      
      
    </div>
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="ccNumber"
    >
      Credit card number
    </label>
    <input
      className="form-control input-custom"
      name="ccNumber"
      onChange={[Function]}
      placeholder="Your 10 digit credit card number"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="cvvNumber"
    >
      CVV number
    </label>
    <input
      className="form-control input-custom"
      name="cvvNumber"
      onChange={[Function]}
      placeholder="Three digit at the back of your card"
      required={true}
      type="password"
      value=""
    />
  </div>
  <div
    className="form-group "
  >
    <label
      htmlFor="expiryDate"
    >
      Expiry date
    </label>
    <input
      className="form-control input-custom"
      name="expiryDate"
      onChange={[Function]}
      placeholder="Your credit card expiry date"
      required={true}
      type="date"
      value=""
    />
  </div>
  <button
    className="btn btn-custom"
    type="submit"
  >
    Submit
  </button>
</form>
`;
